### PR TITLE
New error condition try-except-raise

### PR DESCRIPTION
**Pull request by atodorov**
reports on except handlers where raise is the first operator in the handler body.

There are some failures from existing code which need to be examined. I'm not sure how do we want to handle some scenarios. 

The motivation behind this checker comes from a legacy code base which has a lot of
```
try:
    do_something
except:
    raise
```

which obviously doesn't do anything useful. 

There's also the use case where `raise` is followed by some other nodes which is also detected by unreachable. 

Then there's the scenario where we `raise ExcType()` and the edge case where we raise the same exception type as the one that has been caught in the except clause. 

Let me know how do you like this PR to evolve.

### CODE REVIEW

**Comment by brycepg** [comment]
Looks like your new message is being raised in other functional tests. You'll need to investigate those issues before the tests pass

**Comment by PCManticore** [comment]
@atodorov I would expect this checker to raise only when a bare `raise` is used, which is definitely superfluous. But from what I see, this check is also catching useful constructs, such as when reraising an exception:
```
try:
   ...
except ValueError:
   raise SomeOtherException
```
Regarding your other points, can you point some examples? I'm not sure how prevalent is to raise the same exception, but I'm not opposed to verifying that as well.

**Comment by atodorov** [comment]
@PCManticore looking at the failing test all of the examples are of the type you describe
```
try:
    ...
except SomeException:
    raise AnotherException
```

new checker should not flag this pattern since it is valid use!


A corner case of the above is in Django where
```
try:
    x = Model.objects.get()
except Model.DoesNotExist:
    raise Http404
```

This one we can handle in `pylint_django` and replace with the `get_object_or_404` shortcut.


The edge case I'm talking about is:
```
try:
    int(x)
except ValueError:
    raise ValueError('X is not a proper integer')
```

^^^ this one I do have fairly often with the code base I'm working with. I believe the reason for that was the original author was trying to give overly user-friendly messages. In this example the exception type 
in the handler and in the raise are the same. 

This is the only example I'm not sure whether or not pylint should flag.

**Comment by atodorov** [comment]
FYI updated the new checker to flag only when there is a bare raise. Let me know what you think about the edge case above so I can do further updates or include changelog and docs updates in this PR.

**Comment by PCManticore** [comment]
@atodorov The case you mentioned, reraising the same value but with a different message, sounds good to me to be in pylint as well! Thanks for tackling this!

File: pylint/test/functional/inconsistent_returns.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#pylint: disable=missing-docstring, no-else-return, invalid-name, unused-variable, superfluous-parens
+#pylint: disable=missing-docstring, no-else-return, invalid-name, unused-variable, superfluous-parens, try-except-raise
```
Comment (PCManticore) [inline pylint/test/functional/inconsistent_returns.py]: Curious that I'm not seeing any occurrence of the new error check in this file. Why was it disabled?

Comment (atodorov) [reply pylint/test/functional/inconsistent_returns.py]: That was probably when the new error check was also raised on non-bare raise. Thanks for catching this. I will update the code.

**Comment by atodorov** [comment]
@PCManticore updated the checker to flag code which matches the corner case described above. Let me know if this looks good and I will update the docs/changelog files.

File: pylint/checkers/exceptions.py [hunk 2]
```diff
@@ -347,17 +353,40 @@ def _check_catching_non_exception(self, handler, exc, part):
...
+                    if isinstance(handler.body[0].exc, astroid.Call):
+                        raise_type = handler.body[0].exc.func
+
+                    # flags only when the exception types of the handler
+                    # and the raise statement match b/c we're raising the same
+                    # type of exception that we're trying to handle. Example:
+                    #
+                    # except ValueError:
+                    #   raise ValueError('some user friendly message')
+                    if isinstance(handler.type, astroid.Name) and \
```
Comment (PCManticore) [inline pylint/checkers/exceptions.py]: We prefer using parens continuation rather than using `\`.

**Comment by pbasista** [comment]
I would like to point out that immediately raising in the `except` block *can* actually be useful. In particular, it might be desirable to reraise an exception for a certain exception subtype, while handling it gracefully in general. For example:

```python
try:
    f()
except UnicodeError:
    # it might be desirable to handle UnicodeErrors upstream
    raise
except ValueError:  # parent class of UnicodeError
    # it might be desirable to handle ValueErrors
    # and derived exceptions other than UnicodeError here
    print("something went wrong")
```
In this example, immediately raising in the `except UnicodeError` block is a simple way to ensure that a certain exception subtype is handled differently than its parent exception type and other derived subtypes, which are handled in the next except block.

Therefore, I believe that Pylint should not blindly warn about every `raise` as the first statement in an except block. Instead, it should for instance examine the exception types handled by the following exception handlers of the same `try` block and report an issue only when the `except` block which has `raise` as the first statement is only handling exceptions of types that are *not* derived from at least one exception type handled by some of the following exception handlers.

The above suggestion is only a vague description of a possible approach, which would need to be crafted into a more profound rule. But the point is that considering *all* `except` blocks with `raise` as their first statement as useless or otherwise bad is in my opinion incorrect.

**Comment by atodorov** [comment]
@pbasista can you provide a more detailed example of how you handle the exceptions and open a new issue for this so it can be discussed in detail.

**Comment by pbasista** [comment]
There is a 4 days old issue describing the problem: https://github.com/PyCQA/pylint/issues/2284. I did not look properly before commenting here, sorry for that.

It also seems like it is being addressed already by this PR: https://github.com/PyCQA/pylint/pull/2308
